### ISSUE TITLE
Use CUDA graphs both in PyTorch (same as XLA does)

### ISSUE THREAD

**Issue by lezcano** [haifeng-jin/keras-benchmarks#1]
At the moment, these benchmarks are not as realistic as they could be. Here are a couple things:

- [x] XLA-backends are using CUDA graphs, but PyTorch is not. Tu run PyTorch with CUDA graphs, pass `torch.compile(mode="reduce-overhead")`. Even more, if you are looking for the best-performing option, you should pass `torch.compile(mode="max-autotune")`. When you do this, consider running the model twice as a warm-up, as PyTorch uses the first two executions to collect the necessary metadata to create the cuda-graphs.
- [x] The loss function is not being compiled. It should be compiled together with the whole model.

As a reality check, you can compare your running times against https://hud.pytorch.org/benchmark/compilers see if you can roughly reproduce the results there with your set-up (provided you are benchmarking on A100s). This reports the running time for certain configurations of training/inference / differnt dtypes. You'll still need to look at the batchsize / optimiser used in https://github.com/pytorch/benchmark/.

**Comment by haifeng-jin** [comment]
@lezcano Thanks for the issue!
I will let you know if the number changes after enabling cuda graphs.

**Comment by haifeng-jin** [comment]
We are now using reduce-overhead mode and compiling the loss function.
We got a significant boost of the BERT performance.
It is as fast as Keras 3 for training and Keras 3 is now ~75% faster than native PyTorch for inferencing.
I had to reduce the batch size since the compilation is taking more memory.

The rest of the models has remained the same.
Mistral fit and Gemma fit compiled are much slower (~4x) than not compiled, so we did not use compilation for these two.

The blog post will be updated soon.

Thank you so much again for the suggestions!

**Comment by haifeng-jin** [comment]
Blog post has been updated.

**Comment by lezcano** [comment]
Regarding compiling the loss function, you should compile it together with the model. See https://github.com/haifeng-jin/keras-benchmarks/commit/79d74e509e92aec0ee0913482d4eee2158093529#r139861596

**Comment by haifeng-jin** [comment]
Thanks!
Reopening the issue since sam training and stable diffusion training may still have some potential for improvement.

**Comment by haifeng-jin** [comment]
Stable diffusion training has a significant improve ~20% after this change.
The rest are similar.
As we are compiling the entire functions, I also changed how we use it in sam inference. Also saw some improvement there.
Had to change the batch size for it since it runs OOM with the old batch size.

Blog post updates in progress.

**Comment by haifeng-jin** [comment]
Blog post updated. Will show on the website at the next  push.
